packages/adblocker-electron/adblocker.ts

```diff
@@ -7,17 +27,125 @@ export function fromElectronDetails({
   url,
   resourceType,
   referrer,
-}: {
-  url: string;
-  resourceType: ElectronRequestType;
-  referrer: string;
-}): Request {
+}: Electron.OnBeforeRequestDetails | Electron.OnHeadersReceivedDetails): Request {
   return Request.fromRawDetails({
     sourceUrl: referrer,
-    type: resourceType || 'other',
+    type: (resourceType || 'other') as ElectronRequestType,
     url,
   });
 }
 
+/**
+ * Wrap `FiltersEngine` into a Electron-friendly helper class.
+ */
+export class ElectronBlocker extends FiltersEngine {
+  public enableBlockingInSession(ses: Electron.Session) {
+    ses.webRequest.onHeadersReceived({ urls: ['<all_urls>'] }, this.onHeadersReceived);
+    ses.webRequest.onBeforeRequest({ urls: ['<all_urls>'] }, this.onBeforeRequest);
+    ipcMain.on('get-cosmetic-filters', this.onGetCosmeticFilters);
+    ses.setPreloads([join(__dirname, './content.js')]);
+  }
+
+  private onGetCosmeticFilters = (
+    e: Electron.IpcMessageEvent,
+    url: string,
+    id: string,
+    msg: IBackgroundCallback & { action?: string },
+  ): void => {
+    // Extract hostname from sender's URL
+    const parsed = parse(url);
+    const hostname = parsed.hostname || '';
+    const domain = parsed.domain || '';
+
+    const { active, styles, scripts } = this.getCosmeticsFilters({
+      domain,
+      hostname,
+      url,
+
+      classes: msg.classes,
+      hrefs: msg.hrefs,
+      ids: msg.ids,
+
+      // This needs to be done only once per frame
+      getBaseRules: msg.lifecycle === 'start',
+      getInjectionRules: msg.lifecycle === 'start',
+      getRulesFromHostname: msg.lifecycle === 'start',
+
+      // This will be done every time we get information about DOM mutation
+      getRulesFromDOM: msg.lifecycle === 'dom-update',
+    });
+
+    if (active === false) {
+      return;
+    }
+
+    // Inject custom stylesheets
+    this.injectStyles(e.sender, styles);
```

Comment:
Wondering if it would be technically possible to inject both styles and scripts from either the main process or the renderer? Instead of splitting them.

Reply:
Unfortunately it breaks up when I moved injecting scripts to main process. In renderer process there's no `insertCSS` method in `webFrame`.